### DISCUSSION TITLE
Changing the Lutris Community Installer UI

### DISCUSSION THREAD

**Discussion by danieljohnson2** [lutris/lutris#3998] [Ideas]
When you switch Lutris to the "Lutris" sidebar row, you get a widget to select between "Your Library" and "Community Installers"- the later showing your library as known to Lutris.net. The later lets you search the entire Lutris website. Both are views on the Lutris.net website.

It looks like this:

![Screenshot from 2022-01-17 09-08-15](https://user-images.githubusercontent.com/6507403/149785624-ab85c696-2b69-49f0-9487-bba402e12eb7.png)

And that's kind of strange. Kind of like a stack switcher, kind of like a notebook. I'm not sure what the intent was, but it looks.. unpleasant. The buttons are linked, but not fully squared off. Look at "Your Library", zoomed in:

![Screenshot from 2022-01-17 09-08-37x2](https://user-images.githubusercontent.com/6507403/149785822-7d2f63b3-4962-477e-a4a4-f96c1e590157.png)

See the gaps in the corners? That's the space between the rounded edges of the button, and the sidebar and header bar next to it. Not very classy. What would be better?

Well, if we're going to be GNOME-y, we could stuff everything into the header-bar. Abusing title-bars is what makes GNOME GNOME, right? This is dead easy, but you really need to use icons here or it is much too wide. Like this:

![Screenshot from 2022-01-17 09-07-56](https://user-images.githubusercontent.com/6507403/149786174-d223e190-db72-4672-84e7-f1e3a4513fa5.png)

This frees up vertical space, so there's more space for games here. But the icons are just what I could find among the standard icons known to Glade, and are a bit mysterious. Tool-tips can help, and maybe custom designed icons would too.

But the position of the buttons now suggests to me that maybe a notebook was intended. Like this:

![Screenshot from 2022-01-17 09-07-28](https://user-images.githubusercontent.com/6507403/149786136-cf1e085d-bc55-4e47-b415-de3007256093.png)

That's a standard GTK notebook jammed in there with a crowbar- it's going to be much more difficult to make this actually *work*. It also uses slightly *more* vertical space than what we have now. But it looks very nice, and if the intent was a notebook like this, well, this is a notebook.

The final option that occurs to me is to just square off the buttons, with styles. Not too hard. Like so:

![Screenshot from 2022-01-17 09-44-54](https://user-images.githubusercontent.com/6507403/149790796-0a5dbf28-ce83-49f8-bac1-9ecd6218f16a.png)

This removes some of the button borders and zeros the border-radius. Just like what we have now, but neater looking. But it still does not look like the standard GTK controls you'd use for this.

Me, on balance I like the first option (use the header-bar)- this isn't *really* a notebook in that there is only one list here, only one set of underlying data- the switch here is how its filtered. Also it's easy, and header-bars are fun.

But I'm talking about going beyond a bug fix here, and this isn't a clearly needed feature. I don't want to just pop a PR and hope. I'd like to get other people's views. Maybe other ideas for this.

I'm particularly interested in @strycore's input. He might know what the original intent was here, where I'm just guessing at it.

**Comment by AlexanderRavenheart** [comment]
My vote is for the first option, as that is exactly what headerbars were made for, to optimize available space and use iconography to be slim and efficiemt instead of having bulky wordy buttons.

**Reply by danieljohnson2** [reply]
Of course, this means using tiny inscrutable pictures. What do you think of my tiny inscrutable picture choices? For "Your Library" I used the same icon as the "Games" sidebar row, and for "Community Installers" I used an icon for 'network server'.

**Comment by strycore** [comment]
The plans is to remove the tabs because this whole thing is just a hack.
The search for community installers ended up here because it was the easiest way to add it to Lutris but it's not where's it's supposed to be. 
The Lutris tab should display your Lutris library, that's it. That's what other services show. 
The search for installer should be done in the "Add game" dialog. Not the one we have at the moment but a new one that lets users search for installers / import games / scan folders / etc.

**Reply by danieljohnson2** [reply]
I guess that make sense. Is this plan for the next Lutris release?

If so, my time might be better spent on a new 'Add Game' dialog showing this content.

**Reply by strycore** [reply]
I would like to have it done as soon as we can. With our recent import of IGDB games, the "community installers" tab will be even more confusing. We also support a new image format which won't work with this tab either. 

One aspect of the new dialog is that searches don't load any banner or icons, so the whole process is a lot lighter.

**Reply by danieljohnson2** [reply]
I could do that. Though I favor an approach where banners and icons do load, but only the ones that are visible.

But you're the boss. Maybe getting banners for all the games that have recently turned up on Lutris.net is impractical.

What's the new image format? I see you've added 'cover-art', but that seems like it ought to work.

**Reply by strycore** [reply]
For searching through our database, it will be much more useful to show additional metadata like year, platforms, genres and developer in the search result. This is data we get in the API response, there is no additional media to download. We can start downloading the media when the user has selected a game from the search results. That's at most 3 requests (if we download icon, banner and cover) vs 200 requests of whatever format is being displayed.

The reason covers don't work for search is that we don't have predictable URLs for them. That's also why we only show the covers for the games that are in a user's library.

**Reply by danieljohnson2** [reply]
I'm thinking of something like this:

![Screenshot from 2022-01-18 16-57-23](https://user-images.githubusercontent.com/6507403/150025661-2d9c9fdc-282a-40e2-9db6-441d5e27695c.png)

Though without the icons and with a requirement to search before the list populates. (I threw this together with existing parts, so it's not quite right). Select the game, then click install to go to the install window. This dialog is modal, but the install window is not.

It's meant to look like the 'open file' dialog. But I notice that Lutris avoids using header-bars like this in dialogs (only in the main window and the log window). I'd rather use the header-bar: GNOME isn't properly GNOME-y without a silly red hat.

Any feedback?

**Comment by strycore** [comment]
Absolutely *not*!

**Reply by danieljohnson2** [reply]
No feedback? :slightly_frowning_face:

Well, fine, I'll find out if you like it from the fate of the eventual PR.

But be warned: if it fails just *because* you wouldn't give any feedback, I will go back in time to 2015 and put a silly GNOME-y hat on your file browser. Don't think I wouldn't!

**Comment by danieljohnson2** [comment]
OK, I've got an update on this effort and a question for @strycore, when you get a chance.

Here's where I've got to:
![Screenshot from 2022-01-20 19-23-45](https://user-images.githubusercontent.com/6507403/150443522-a083899a-41a7-44a4-88c6-02447beb23c5.png)

You get three install commands here: "Add Other..." triggers the same add-game dialog we had before, for the manual experience. "Install" fires off the install window. "Locate Installed Game" does the add-game dialog again, but preloaded with the data from Lutris.net.
 
I've removed the icon column here, but also many others. The name and year survive, but that's it. That's because that seems to be all the game meta-data I am getting. No platform, no runner. Of course no "last played" or "installed at".

But I expected to get more. The Lutris.net website gives me platforms and genres and publishers in my web-browser. but I don't see this in the API result. And stock Lutris doesn't display this stuff in the "Community Installers" tab either.

So my question is: Am I missing something here, or this just not supposed to be there?

**Comment by strycore** [comment]
What I had in mind puts way less emphasis on having a list of games.  I want an initial screen that gives a list of different ways to add games to Lutris, that would allow:

* Searching for games, I believe a ComboBox should be enough to display the results. Stuff like platform, year could be added on a second line with a smaller font. This might be an issue for some games from the 80s and 90s which got released for a massive number of systems.
* Scan a folder for ROMs. We already have some initial code for scanners which I've used to add some ROMs to RetroArch.
* Scan a folder for previously installed Lutris games. (can possibly be integrated in the scanners option in the GUI)
* Manually installing a game (from a CD / setup.exe / etc...)
* Manually add an already installed game

So the window wouldn't look like a list of games, it would look like a selection screen with 4 or 5 different options to choose different install methods.

**Comment by strycore** [comment]
Not a final design, but something along those lines

![20220120_205058](https://user-images.githubusercontent.com/280336/150468044-66382800-0163-4ad5-b9fc-a7c16ed88e2c.jpg)

**Comment by strycore** [comment]
Something to note: We use dropdown for the Lutris internal dashboard for the game merge tool. This works well enough but we need more metadata displayed to make it easier to differentiate between releases.  Most important piece of data is the year, then the systems the game was released on, then the publisher and developer.

**Comment by danieljohnson2** [comment]
That's quite different, and makes Lutris.net much less central.

I can dig it.

I don't think a traditional combo-box is really an adequate way to list matching games, though, but I could imagine some some other sort of pop-up thing. Better than a wizard, I think.

However, the api is still not giving me the platform, so I can't show that. Or if I can, I don't know how.

**Comment by strycore** [comment]
The goal is really to make the Lutris as independent from lutris.net as possible. It should be able to install a vast majority of games even without access to the website.
Previously, we needed install scripts for every game. Now, I really want to reserve install scripts for edge cases that lutris can't figure out by itself.

**Comment by danieljohnson2** [comment]
I've done some experiments with your add-games dialog proposal:
![Screenshot from 2022-01-22 08-50-05](https://user-images.githubusercontent.com/6507403/150641089-0628b2f6-d693-4f85-ae66-118cf73d9288.png)

If you use the search box, you get this:
![Screenshot from 2022-01-22 09-01-36](https://user-images.githubusercontent.com/6507403/150641448-cc955ea5-86b9-40cc-890f-57174ae68994.png)

The normal games list doesn't fit at all, but since I have only name and year to show I just slapped together a list-box. The pop-over can't extend outside the window due to GTK limitations (except on Wayland), so I made the dialog a bit too big.

But it leaves lots of space in the main dialog, and still lets us reveal the 'install' and 'locate' buttons in a way that seems intuitive.

Is this in line with your thoughts?

It occurs to me that it might be better to display this result list inline in the dialog, as an extra (large) row below the search, displacing the "manual" option downwards instead of cover it with a pop-over. I'm not too sure how important it is to you that this be like a combo-box.

**Comment by strycore** [comment]
This type of widget does not exist. No software does that and I'm not going to let this happen in Lutris.

* The "Locate installed game" goes on the next step, the one where you choose which installer to launch.
* The year is not in a smaller or greyed font, it's just italic
* The metadata line isn't always present making list elements visually inconsistent.
* We have examples of code that uses a dropdown for a very large list of elements (MAME platforms)
* "The normal games list doesn't fit at all" there is no game list, the goal is to find a *single* game.

**Comment by danieljohnson2** [comment]
Hey, no offense intended! I'm just trying to help, not to derail your vision of Lutris.

But I think it's clear enough that I can't help you this way. I'll leave this area to you, to handle your own way.

I seem to do better with more complete, more focused, and more polished PRs that, well, don't look like half-finished ass.
